### PR TITLE
 chore: sanity test meta module via standalone devimint test

### PR DESCRIPTION
**Pull request by dpc**
I thought it would be good to have basic test for meta module, and I discovered a problem.

To reproduce:

```
env RUST_LOG=server::module::meta=trace,info ./scripts/tests/meta-module-test.sh
```

Sometimes (quite often on my system) it will fail timing out polling for reaching a consensus for a meta value.

You can look up the `/tmp/devimint...` dir and look at the peer logs (e.g. `less /tmp/devimint-3273235-123/logs/fedimintd-3.log`)

This is from a fedimintd-3, that seemed to fail to submit:

```
2024-03-29T18:12:41.445613Z DEBUG server::module::meta: Received a submitted a value peer_id=1 key=0 value=7b22666f6f223a22626172227d salt=1
3548918736504005055
2024-03-29T18:12:41.445673Z  INFO server::module::meta: Peer submitted a value peer_id=1 key=0 value_len=13 matching=1 threshold=3
2024-03-29T18:12:41.771767Z  INFO server::module::meta: Our own guardian submitted a value key=0 value_len=13
2024-03-29T18:12:42.096532Z DEBUG server::module::meta: Received a submitted a value peer_id=2 key=0 value=7b22666f6f223a22626172227d salt=1
196510382307997764
2024-03-29T18:12:42.096603Z  INFO server::module::meta: Peer submitted a value peer_id=2 key=0 value_len=13 matching=2 threshold=3
2024-03-29T18:12:42.748009Z DEBUG server::module::meta: Received a submitted a value peer_id=0 key=0 value=7b22626172223a2262617a227d salt=1
5463300193985426068
2024-03-29T18:12:42.748090Z  INFO server::module::meta: Peer submitted a value peer_id=0 key=0 value_len=13 matching=1 threshold=3
```

The `Our own guardian submitted a value` is followed by receiving that submission in `process_consensus_item` with `peer_id=2` instead of `3`!

I was under impression that `peer_id` will always mean originator of the message, but it doesn't look like that's the case. If indeed it is not guaranteed, then we should not be passing it to `process_consensus_item` as it doesn't really mean anything reliable on this level of abstraction, and it will just lead to bugs like this one. And I'll need to add some signing etc.

Or maybe it should be reliable, and it's a bug somewhere.

Please sanity check my findings and let me know.

### CODE REVIEW

**Comment by dpc** [comment]
Maybe I jumped to conclusions. After getting back to it and comparing logs between peers it seems more like just Peer-Id=3 dropped its message:

`/tmp/devimint-3273235-123/logs/fedimintd-0.log`

```
2024-03-29T18:12:41.445217Z DEBUG server::module::meta: Received a submitted a value peer_id=1 key=0 va
lue=7b22666f6f223a22626172227d salt=13548918736504005055
2024-03-29T18:12:41.445301Z  INFO server::module::meta: Peer submitted a value peer_id=1 key=0 value_le
n=13 matching=1 threshold=3
2024-03-29T18:12:41.570589Z  INFO server::module::meta: Our own guardian submitted a value key=0 value_
len=13
2024-03-29T18:12:42.096105Z DEBUG server::module::meta: Received a submitted a value peer_id=2 key=0 va
lue=7b22666f6f223a22626172227d salt=1196510382307997764
2024-03-29T18:12:42.096186Z  INFO server::module::meta: Peer submitted a value peer_id=2 key=0 value_le
n=13 matching=2 threshold=3
2024-03-29T18:12:42.747472Z DEBUG server::module::meta: Received a submitted a value peer_id=0 key=0 va
lue=7b22626172223a2262617a227d salt=15463300193985426068
2024-03-29T18:12:42.747564Z  INFO server::module::meta: Peer submitted a value peer_id=0 key=0 value_le
n=13 matching=1 threshold=3
```

`/tmp/devimint-3273235-123/logs/fedimintd-1.log`

```
2024-03-29T18:12:41.023865Z  INFO server::module::meta: Our own guardian submitted a value key=0 value_
len=13
2024-03-29T18:12:41.445703Z DEBUG server::module::meta: Received a submitted a value peer_id=1 key=0 va
lue=7b22666f6f223a22626172227d salt=13548918736504005055
2024-03-29T18:12:41.445787Z  INFO server::module::meta: Peer submitted a value peer_id=1 key=0 value_le
n=13 matching=1 threshold=3
2024-03-29T18:12:42.096777Z DEBUG server::module::meta: Received a submitted a value peer_id=2 key=0 va
lue=7b22666f6f223a22626172227d salt=1196510382307997764
2024-03-29T18:12:42.096876Z  INFO server::module::meta: Peer submitted a value peer_id=2 key=0 value_le
n=13 matching=2 threshold=3
2024-03-29T18:12:42.748077Z DEBUG server::module::meta: Received a submitted a value peer_id=0 key=0 va
lue=7b22626172223a2262617a227d salt=15463300193985426068
2024-03-29T18:12:42.748179Z  INFO server::module::meta: Peer submitted a value peer_id=0 key=0 value_le
n=13 matching=1 threshold=3
```

`/tmp/devimint-3273235-123/logs/fedimintd-2.log:`

```
2024-03-29T18:12:41.445632Z DEBUG server::module::meta: Received a submitted a value peer_id=1 key=0 va
lue=7b22666f6f223a22626172227d salt=13548918736504005055
2024-03-29T18:12:41.445696Z  INFO server::module::meta: Peer submitted a value peer_id=1 key=0 value_le
n=13 matching=1 threshold=3
2024-03-29T18:12:41.670564Z  INFO server::module::meta: Our own guardian submitted a value key=0 value_
len=13
2024-03-29T18:12:42.096582Z DEBUG server::module::meta: Received a submitted a value peer_id=2 key=0 va
lue=7b22666f6f223a22626172227d salt=1196510382307997764
2024-03-29T18:12:42.096645Z  INFO server::module::meta: Peer submitted a value peer_id=2 key=0 value_le
n=13 matching=2 threshold=3
2024-03-29T18:12:42.747842Z DEBUG server::module::meta: Received a submitted a value peer_id=0 key=0 va
lue=7b22626172223a2262617a227d salt=15463300193985426068
2024-03-29T18:12:42.747930Z  INFO server::module::meta: Peer submitted a value peer_id=0 key=0 value_le
n=13 matching=1 threshold=3
```


`/tmp/devimint-3273235-123/logs/fedimintd-3.log`

```
2024-03-29T18:12:41.445613Z DEBUG server::module::meta: Received a submitted a value peer_id=1 key=0 va
lue=7b22666f6f223a22626172227d salt=13548918736504005055
2024-03-29T18:12:41.445673Z  INFO server::module::meta: Peer submitted a value peer_id=1 key=0 value_le
n=13 matching=1 threshold=3
2024-03-29T18:12:41.771767Z  INFO server::module::meta: Our own guardian submitted a value key=0 value_
len=13
2024-03-29T18:12:42.096532Z DEBUG server::module::meta: Received a submitted a value peer_id=2 key=0 va
lue=7b22666f6f223a22626172227d salt=1196510382307997764
2024-03-29T18:12:42.096603Z  INFO server::module::meta: Peer submitted a value peer_id=2 key=0 value_le
n=13 matching=2 threshold=3
2024-03-29T18:12:42.748009Z DEBUG server::module::meta: Received a submitted a value peer_id=0 key=0 va
lue=7b22626172223a2262617a227d salt=15463300193985426068
2024-03-29T18:12:42.748090Z  INFO server::module::meta: Peer submitted a value peer_id=0 key=0 value_le
n=13 matching=1 threshold=3
```

**Comment by dpc** [comment]
Damn, it seems like log file numbers don't correspond to peer-ids ... :D

**Comment by dpc** [comment]
BTW. Sometimes something else gets stuck, but it's less common, so I'm focused on the common problem.

I verified that the peer-3 does return right proposal from  `fn consensus_proposal` - it just doesn't seem to get delivered to any peer (including peer-3), and peer-3 keeps returning it from that function.

**Comment by dpc** [comment]
@joschisan

**Comment by elsirion** [comment]
Pretty sure this is racy:

https://github.com/fedimint/fedimint/blob/5918169d58d10d5f2febcb48cab1ddf36946a210/modules/fedimint-meta-server/src/lib.rs#L231-L236

Submitting something to consensus doesn't guarantee inclusion, so unless a "best effort" approach is sufficient here we can't delete the key until our submission was accepted.

**Comment by dpc** [comment]
> Submitting something to consensus doesn't guarantee inclusion, so unless a "best effort" approach is sufficient here we can't delete the key until our submission was accepted.

```
 } else if Self::get_consensus(dbtx, key).await.as_ref() == Some(&value) { 
```

is only true if the value is already part of a consensus.

The value is actually deleted only when a new value (not necessarily for the same key) is being submitted via API.

So, yes, it's technically a race, but it will only happen if the guardian submits values to different keys in quick succession, and the guardian was trying to clear their own submission that did not become a consensus to be exactly like consensus value. The race will cause the value not to be cleared, which does not matter for anything but ... aesthetic purposes (other guardians can still see that you voted for something different).

But now that I think about again,  I guess it might be easily fixable...

@elsirion

**Comment by dpc** [comment]
Why is it always last peer that fails to deliver... ?

**Comment by dpc** [comment]
We had a call with @joschisan and it seems like we had a case of consensus round lockstep, which can be fixed with: https://github.com/fedimint/fedimint/pull/4846 

Now everything passes consistently.

File: devimint/src/util.rs [hunk 1]
```diff
@@ -438,6 +439,39 @@ where
...
+                        "Polling {name} failed after {attempt} retries (timeout: {}s)",
+                        timeout.as_secs()
+                    )
+                });
+            }
+        }
+    }
+
+    unreachable!();
+}
```
Comment (elsirion) [inline devimint/src/util.rs:473]: Can't this be a much shorter wrapper around `poll` where you call `.map_err(ControlFlow::Continue)` on `f`?

**Review by elsirion — COMMENTED** [review]
Looks mostly good

File: devimint/src/util.rs (hunk 1 above)
Comment (maan2003) [reply devimint/src/util.rs:473]: or even the new fancy `fedimint_core::util::retry`

File: fedimint-client/src/sm/executor.rs [hunk 2]
```diff
@@ -736,7 +736,7 @@ impl ExecutorInner {
 
 impl Debug for ExecutorInner {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        let (active, inactive) = futures::executor::block_on(async {
+        let (active, inactive) = tokio::runtime::Handle::current().block_on(async {
```
Comment (maan2003) [inline fedimint-client/src/sm/executor.rs:739]: WASM!!!!!

File: fedimint-client/Cargo.toml [hunk 3]
```diff
@@ -39,7 +39,7 @@ serde_json = { workspace = true }
 strum = { workspace = true }
 strum_macros = { workspace = true }
 thiserror = { workspace = true }
-tokio = { version = "1.36.0", features = [ "time", "macros" ] }
+tokio = { version = "1.36.0", features = [ "time", "macros", "rt" ] }
```
Comment (maan2003) [inline fedimint-client/Cargo.toml:42]: tokio has runtime on wasm :eyes: does it work?

File: .config/semgrep.yaml [hunk 4]
```diff
@@ -72,6 +72,13 @@ rules:
   pattern: tokio::time::sleep
   severity: WARNING
 
+- id: ban-futures-block-on
+  languages:
+    - rust
+  message: Use `tokio::runtime::Handle::current.block_on()` instead.
+  pattern: futures::executor::block_on
```
Comment (maan2003) [inline .config/semgrep.yaml:79]: finally
was this real cause of bug

Comment (dpc) [reply .config/semgrep.yaml:79]: Nope.

File: devimint/src/util.rs (hunk 1 above)
Comment (dpc) [reply devimint/src/util.rs:473]: When I wrote it non of these were there yet. :D

File: fedimint-client/Cargo.toml (hunk 3 above)
Comment (dpc) [reply fedimint-client/Cargo.toml]: https://github.com/fedimint/fedimint/pull/4839/files landed so i hope so.

File: .config/semgrep.yaml (hunk 4 above)
Comment (dpc) [reply .config/semgrep.yaml]: https://github.com/fedimint/fedimint/pull/4846 did

File: fedimint-client/Cargo.toml (hunk 3 above)
Comment (maan2003) [reply fedimint-client/Cargo.toml]: lets see https://github.com/fedimint/fedimint/pull/4850

File: modules/fedimint-meta-tests/src/bin/meta-sanity-test.rs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+use devimint::util::poll_simple;
+use devimint::{cmd, util};
+use fedimint_core::PeerId;
+use semver::Version;
+use serde_json::json;
+use tracing::{info, warn};
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    devimint::run_devfed_test(|dev_fed| async move {
```
Comment (maan2003) [inline modules/fedimint-meta-tests/src/bin/meta-sanity-test.rs:14]: I would move the closure to a outer function, just to reduce a level of indentation

**Review by maan2003 — APPROVED** [review]
ACK https://github.com/fedimint/fedimint/pull/4772/commits/df14c976613badf13fc05f7c365c06c333ea628c https://github.com/fedimint/fedimint/pull/4772/commits/dbe5aef56f49c365c8a4f8dfa9ce12f212b59044 https://github.com/fedimint/fedimint/pull/4772/commits/aa545c4ecc746bb355ff39d59617bbd0c7fc0675

**Comment by maan2003** [comment]
also nice tests!

**Review by maan2003 — DISMISSED** [review]
ACK all commits but 5e2cfad85510b4b3bc13bf76e9874fe9e2a917a5

I am not familiar with that code

File: modules/fedimint-meta-tests/src/bin/meta-sanity-test.rs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        )
+        .await?;
+
+        // form a consensus with a minority vote
+        submit(&client, PeerId::from(0), &minority_submission_value).await?;
+        submit(&client, PeerId::from(2), &submission_value).await?;
+        assert_eq!(
+            submit(&client, PeerId::from(3), &submission_value).await?,
+            serde_json::Value::Bool(true)
+        );
```
Comment (elsirion) [inline modules/fedimint-meta-tests/src/bin/meta-sanity-test.rs:142]: nit: Does this `true` carry any meaning? Seems to be statically returned.

https://github.com/fedimint/fedimint/blob/fedd08cb69cad763fb3511e15f08e252af5a1ec0/modules/fedimint-meta-client/src/cli.rs#L99-L113

Should we test it for all submissions or maybe none?

**Review by elsirion — DISMISSED** [review]

**Comment by elsirion** [comment]
Want to give @joschisan an opportunity to look at 5e2cfad85510b4b3bc13bf76e9874fe9e2a917a5.

**Comment by dpc** [comment]
Conflict. Rebased.

**Review by elsirion — APPROVED** [review]

**Review by maan2003 — APPROVED** [review]

**Pull request merged**